crates/red_knot_python_semantic/src/types.rs

```diff
@@ -471,6 +471,232 @@ impl<'db> Type<'db> {
+    /// Note: This function aims to have no false positives, but might return
+    /// wrong `false` answers in some cases.
+    pub(crate) fn is_disjoint_from(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        match (self, other) {
+            (Type::Never, _) | (_, Type::Never) => true,
+
+            (Type::Any, _) | (_, Type::Any) => false,
+            (Type::Unknown, _) | (_, Type::Unknown) => false,
+            (Type::Unbound, _) | (_, Type::Unbound) => false,
+            (Type::Todo, _) | (_, Type::Todo) => false,
+
+            (ty @ (Type::Function(..) | Type::Module(..) | Type::Class(..)), other)
+            | (other, ty @ (Type::Function(..) | Type::Module(..) | Type::Class(..))) => {
+                ty != other
+            }
+
+            (Type::Union(union), other) | (other, Type::Union(union)) => union
+                .elements(db)
+                .iter()
+                .all(|e| e.is_disjoint_from(db, other)),
+
+            (Type::Intersection(intersection), other)
+            | (other, Type::Intersection(intersection)) => {
+                if intersection
+                    .positive(db)
+                    .iter()
+                    .any(|p| p.is_disjoint_from(db, other))
+                {
+                    true
+                } else {
+                    // TODO we can do better here. For example:
+                    // X & ~Literal[1] is disjoint from Literal[1]
+                    false
+                }
+            }
+
+            (Type::Tuple(tuple), other) | (other, Type::Tuple(tuple)) => {
+                if let Type::Tuple(other_tuple) = other {
+                    if tuple.len(db) == other_tuple.len(db) {
+                        tuple
+                            .elements(db)
+                            .iter()
+                            .zip(other_tuple.elements(db).iter())
+                            .any(|(e1, e2)| e1.is_disjoint_from(db, *e2))
+                    } else {
+                        true
+                    }
+                } else {
+                    true
+                }
+            }
+
+            (Type::None, other) | (other, Type::None) => match other {
+                Type::Never
+                | Type::Any
+                | Type::Unknown
+                | Type::Unbound
+                | Type::Todo
+                | Type::Function(..)
+                | Type::Module(..)
+                | Type::Class(..)
+                | Type::Union(..)
+                | Type::Intersection(..)
+                | Type::Tuple(..) => {
+                    unreachable!("handled above")
+                }
+                Type::IntLiteral(_)
+                | Type::BooleanLiteral(_)
+                | Type::StringLiteral(..)
+                | Type::LiteralString
+                | Type::BytesLiteral(..) => true,
+                Type::None => false,
+                Type::Instance(class_type) => !class_type.is_known(db, KnownClass::NoneType),
```

Comment:
Is this save? My thinking was that `NoneType` can't be subclassed.

Reply:
`NoneType` is however a subclass of `object`, so `Type::Instance(builtins.object)` is not disjoint from `Type::None`

```pycon
>>> type(None).__mro__
(<class 'NoneType'>, <class 'object'>)
```